Improve performance of is_function

eendebakpt opened the pull request:
For instruments with many parameters (100 - 1000 range) the overhead of qcodes parameters becomes a bottleneck in startup time. In this PR we improve performance of `is_function`:
```
main: 6.65 μs ± 185 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
PR: 668 ns ± 31.3 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```
Testing code:
```
def f(x):
    return None
arg_count=1
%timeit is_function0(f, arg_count) 
```

(there are more bottlenecks, but `is_fuction` is used in creation of every Parameter and is quite heavy)

<!--

Thanks for submitting a pull request against QCoDeS.

To help us effectively merge your pr please consider the following check list.

- [ ] Make sure that the pull request contains a short description of the changes made.
- [ ] If you are submitting a new feature please document it. This can be in the form of inline
      docstrings, an example notebook or restructured text files.
- [ ] Please include automatic tests for the changes made when possible.

Unless your change is a small or trivial fix please add a small changelog entry:

- [ ] Create a file in the docs\changes\newsfragments folder with a short description of the change.

This file should be in the format number.categoryofcontribution. Here the number should either be the number
of the pull request. To get the number of the pull request one must
first the pull request and then subsequently update the number. The category of contribution should be
one of ``breaking``, ``new``, ``improved``, ``new_driver`` ``improved_driver``, ``underthehood``.

If this fixes a known bug reported against QCoDeS:

- [ ] Please include a string in the following form ``closes #xxx`` where ``xxx``` is the number of the bug fixed.

Please have a look at [the contributing guide](https://microsoft.github.io/Qcodes/community/contributing.html)
for more information.

If you are in doubt about any of this please ask and we will be happy to help.

-->

jenshnielsen commented:
Thanks @eendebakpt That seems very useful. In the long term I think we should consider completely removing and simplifying the command class from parameter but until then this seems like a nice improvement. I did not look into why the tests doesn't like this as is

eendebakpt commented:
@microsoft-github-policy-service agree [company="TNO"]

eendebakpt commented:
@microsoft-github-policy-service agree company=TNO

eendebakpt commented:
> @eendebakpt I don't think it's used much. Perhaps we could change the value for `coroutine` argument to `coroutine: bool | None = None` and skip the check if coroutine=None ? Note that I changed the import in #7753 since the function in asyncio was deprecated in python 3.14

Yes, I had a similar idea. Added to the PR!

jenshnielsen reviewed (approved):
@eendebakpt Thanks, looks good. Could you run the precommit hooks and this should be good to merge. The other ci issue is resolved on main so should be fixed if you rebase or merge main into this
